### PR TITLE
feat(frontend): support alter source pause/resume

### PR DESCRIPTION
**Pull request by lyang24**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

**Please do not leave this empty!**

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
Support alter source pause or resume syntax sugar by setting the rate limit to 0 on pause or to cancel rate limit on resume. Please not PAUSE/ RESUME only works when the source haven't configure the rate limit.
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)
- Refer to a related PR or issue link (optional)
https://github.com/risingwavelabs/risingwave/issues/19181

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

Public Preview: Do note that the structure and format of ALTER SOURCE <name> PAUSE/ RESUME is considered unstable, and may change overtime.

After this PR, two new alter source syntax will be supported.

`ALTER SOURCE <name> PAUSE;` to pause processing stream from source.
`ALTER SOURCE <name> RESUME;` to continue processing stream from source.

### CODE REVIEW

File: src/sqlparser/src/parser.rs [hunk 1]
```diff
@@ -3515,9 +3518,17 @@ impl Parser<'_> {
         } else if self.parse_keywords(&[Keyword::SWAP, Keyword::WITH]) {
             let target_source = self.parse_object_name()?;
             AlterSourceOperation::SwapRenameSource { target_source }
+        } else if self.parse_keyword(Keyword::PAUSE) {
+            AlterSourceOperation::SetSourceRateLimit {
+                rate_limit: SOURCE_RATE_LIMIT_PAUSED,
+            }
+        } else if self.parse_keyword(Keyword::UNPAUSE) {
```
[the remark this answers is not in the document]
Comment (lyang24) [reply src/sqlparser/src/parser.rs]: yep fixed it everywhere

**Comment by tabVersion** [comment]
Hi @lyang24 , thanks for your contribution. 

> ALTER SOURCE <name> PAUSE; to pause processing stream from source.
> ALTER SOURCE <name> RESUME; to continue processing stream from source.

I have a question about the semantics here, please consider the following use case.

```
create source s .... ;

alter s set streaming_rate_limit to 100;
alter s pause; -- source stopped
alter s resume; -- source rate_limit set to unlimited rather than 100
```

Setting rate_limit here may abuse the pause/resume semantic.

cc @fuyufjh, what do you think of the proposed syntax?

**Comment by kwannoel** [comment]
> Hi @lyang24 , thanks for your contribution.
> 
> > ALTER SOURCE  PAUSE; to pause processing stream from source.
> > ALTER SOURCE  RESUME; to continue processing stream from source.
> 
> I have a question about the semantics here, please consider the following use case.
> 
> ```
> create source s .... ;
> 
> alter s set streaming_rate_limit to 100;
> alter s pause; -- source stopped
> alter s resume; -- source rate_limit set to unlimited rather than 100
> ```
> 
> Setting rate_limit here may abuse the pause/resume semantic.
> 
> cc @fuyufjh, what do you think of the proposed syntax?

We can ban `alter source pause` if source already has rate limit set. Likewise, `alter source resume` can be banned, unless rate limit is 0.

**Comment by lyang24** [comment]
> > Hi @lyang24 , thanks for your contribution.
> > > ALTER SOURCE  PAUSE; to pause processing stream from source.
> > > ALTER SOURCE  RESUME; to continue processing stream from source.
> > 
> > 
> > I have a question about the semantics here, please consider the following use case.
> > ```
> > create source s .... ;
> > 
> > alter s set streaming_rate_limit to 100;
> > alter s pause; -- source stopped
> > alter s resume; -- source rate_limit set to unlimited rather than 100
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Setting rate_limit here may abuse the pause/resume semantic.
> > cc @fuyufjh, what do you think of the proposed syntax?
> 
> We can ban `alter source pause` if source already has rate limit set. Likewise, `alter source resume` can be banned, unless rate limit is 0.

The challenge is - The sql parser doesn't know the current rate limit nor can it distinguish from pause or set source_rate_limit to 0. To achieve this we probably need to modify  `SetSourceRateLimit { rate_limit: i32 },` to something like `SetSourceRateLimit { rate_limit: i32 , paused: bool},` and handle in the `handle_alter_streaming_rate_limit`. let me know if there are better approaches.

**Comment by tabVersion** [comment]
> @kwannoel  We can ban alter source pause if source already has rate limit set. Likewise, alter source resume can be banned, unless rate limit is 0.

I am up for this idea.

**Review by kwannoel — APPROVED** [review]
Rest LGTM, thanks for your work!

File: src/frontend/src/handler/alter_streaming_rate_limit.rs [hunk 2]
```diff
@@ -56,6 +58,17 @@ pub async fn handle_alter_streaming_rate_limit(
             let reader = session.env().catalog_reader().read_guard();
             let (source, schema_name) =
                 reader.get_source_by_name(db_name, schema_path, &real_table_name)?;
+            if let Some(prev_limit) = source.rate_limit {
+                if rate_limit == SOURCE_RATE_LIMIT_PAUSED
+                    || (prev_limit != 0 && rate_limit == SOURCE_RATE_LIMIT_RESUMED)
```
Comment (lyang24) [inline src/frontend/src/handler/alter_streaming_rate_limit.rs:63]: this part is quite tricky as we need to able to resume from a paused state where the rate limit is 0

File: e2e_test/source_inline/kafka/alter/resume_pause_source_kafka.slt.serial [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+order by name;
+----
+rl_mv1 SOURCE {SOURCE} 0
+rl_mv2 SOURCE {SOURCE} 0
+rl_mv3 SOURCE {SOURCE} 0
+
+
+skipif in-memory
+statement count 0
+alter source kafka_source resume;
```
[the remark this answers is not in the document]
Comment (lyang24) [reply e2e_test/source_inline/kafka/alter/resume_pause_source_kafka.slt.serial:90]: yep added kafka_source2 with pre configured rate limit, my local risedev d full is having problems, using ci to debug for now. hope it passes.

File: src/frontend/src/handler/alter_streaming_rate_limit.rs [hunk 4]
```diff
@@ -56,6 +58,17 @@ pub async fn handle_alter_streaming_rate_limit(
...
+                {
+                    return Err(InvalidInputSyntax(
+                        "PAUSE or RESUME is invalid when the stream has pre configured ratelimit."
```
Comment (fuyufjh) [inline src/frontend/src/handler/alter_streaming_rate_limit.rs:66]: "PAUSE or RESUME is invalid when the stream has pre configured ratelimit. Just use `SET source_rate_limit = 0`"

Comment (MrCroxx) [reply src/frontend/src/handler/alter_streaming_rate_limit.rs:66]: Will be fixed via #19857

**Comment by MrCroxx** [comment]
Meet the problem that is related to this PR in #19857 .

Using magic numbers for rate limit is kind of tricky to me since there are already various experisions for rate limit. Besides, all of other components like MVs, backfillings, sinks support pause/resume while the rate limit is set.

I have 2 ideas to resolve the problem:

1. Add a `pause` column to the source catalog.
2. Still use 0 as pause and do not rollback the rate limit after resume.

**Pull request merged**